### PR TITLE
Encapsulate SCSS-generated CSS with each component

### LINKED ISSUE #115: Encapsulate SCSS-generated CSS with each component
**Is your feature request related to a problem? Please describe.**

See https://github.com/CityOfDetroit/COD-Design-System/pull/112#issuecomment-1810363230.

**Describe the solution you'd like**

The build script `compile-bootstrap.sh` should be renamed to `compile-styles.sh` and there should be a section in the script for each component with SCSS code to compile. The script should generate a separate CSS per component.

**Describe alternatives you've considered**

See https://github.com/CityOfDetroit/COD-Design-System/pull/112#issuecomment-1810363230 where we currently store all SCSS-generated CSS into a single `themed-bootstrap.css` file.

### PR DESCRIPTION
**Pull request by sreidthomas**
### **Problem**
In PRs 112 and 128, specific components' CSS were added to the customized bootstrap CSS in themed-bootstrap.scss. This approach resulted in the compiled CSS (themed-bootstrap.css) including styles that not all components required. Additionally, importing this CSS into all components increased individual file sizes and posed the risk of conflicting styles among different components utilizing similar CSS selectors.

### **Proposed Solution**
To resolve this issue and enhance maintainability and performance:

**1. Separate Component-Specific CSS**

- Rename compile-bootstrap.sh to compile-styles.sh.

- Develop a function within compile-styles.sh to compile SCSS for each component with custom SCSS, placing the output CSS into the respective component directory.

- Remove component SCSS imports from themed-bootstrap.scss.

**2. Update Component JavaScript**

- Update each component's JavaScript to include the generated CSS for that specific component.

### **Changes Proposed**

**1. Renaming the Compilation Script**

- Renaming compile-bootstrap.sh to compile-styles.sh for clearer functionality.

**2. Separating Component-Specific CSS**

- Implement a function in compile-styles.sh to handle component-wise SCSS compilation.

- Remove component-specific SCSS imports from themed-bootstrap.scss.

**3. Component Integration**

- Update individual components' JavaScript to integrate the generated CSS for each component (**No longer needed because of new proposed changes listed below**)

### BREAK DOWN OF TASKS TO COMPLETE SOLUTION:

- [x] TASK 1: Set up the design system locally -> set it up locally using https://github.com/CityOfDetroit/COD-Design-System#contributing

- [x] TASK 2: Re-name the file **compile-bootstrap.sh** to **compile-styles.sh** in the design system directory after setting it up locally -> file name is **compile_bootstrap.sh** so should be renamed to **compile_styles.sh**

- [x] TASK 3: Add a function in the **compile-styles.sh** to compile SCSS for each component that has custom SCSS. Components ->

**CHECK EACH COMPONENT OFF AS ITS COMPLETED:**
- [x]  table-header
- [x]  table-row
- [x] table-cell-header
- [x]  table-cell
- [x]  nav
- [x] off canvas
- [x]  offcanvasheader
- [x]  offcanvasbody

**MORE NOTES ON TASK 3:** Each component should have its own CSS file. The output CSS should be placed in the relevant component directory.

**NOTE:** The function **compile_sass** already exist as well as the CSS files for each component. 

### **NEW PROPOSED CHANGES:**
Rename **compile_sass** to be called **compile_bootstrap**, then create a new function named **compile_components** that handles compiling each components SCSS to CSS.

For the preexisting CSS files in each components folder: move all the CSS rules in the existing CSS file (if there are any) to the top of the similarly named SCSS file. Then, delete that existing CSS file and let it get generated by SASS and the bash script.

- [x] Re-name **compile_sass** to **compile_bootstrap**

- [x] Move the CSS rules for each component to the top of the similarly named SCSS.file. Then delete those files

 For example with TableHeader:

Move the rules in TableHeader.css into _table-header.scss.
Delete TableHeader.css

In the bash script, make the output of yarn sass be TableHeader.css for the TableHeader component (like yarn sass ${PROJECT_ROOT}src/components/atoms/_table_header.scss ${PROJECT_ROOT}src/components/atoms/TableHeader.css)

- [x] Create a new function named **compile_components**

**With this approach, you also won't have to update the Javascript files for the components, since they're already importing those CSS files (TASK 5 NOT NEEDED)**

- [x] TASK 4: Remove the component SCSS imports from themed-bootstrap.scss -> src -> scss -> themed-bootstrap.scss

- [x] TASK 5: Update each component's Javscript to include the generated CSS for that component (e.g. [like this](https://github.com/CityOfDetroit/COD-Design-System/blob/6a880918ff718f678cfd34798672a37fcb8dab67/src/components/organisms/Table/Table.js#L1C1-L1C47)).
(**UPDATE**: Because of the new proposed changes (**up top**) this task is no longer needed)

### CODE REVIEW

**Comment by sreidthomas** [comment]
One of the test failed after my last commit which consisted of moving the CSS to the respective SCSS files and then deleting the CSS files. Was I supposed to wait to do that? @maxatdetroit 

failedaftermovingCSS

**Comment by maxatdetroit** [comment]
@sreidthomas a good way to test:

1. Run the CSS compilation script: `./scripts/compile-styles.sh`
2. Confirm that the style rules for individual components (e.g. table header) are not present in `src/shared/themed-bootstrap.css`
3. Confirm that the style rules for individual components are present in individual style sheet files in each components directory (e.g. `src/components/atoms/TableHeader.css`).
4. For each component whose stylesheets were moved, view that component in storybook on your local computer (attach screenshots to the PR). Note: Some of the components don't have their own story so you'll have to view the story for the higher level component that uses it. For example, there's no story for table header, so you'll have to view the story for Table and ensure the header looks right (compared to the `dev` branch).


If you notice something is wrong at any of those steps, you can always revert the generated files by doing `git restore X.css` where `X.css` is the file you want to revert.

**Comment by sreidthomas** [comment]
The test that just failed would that stop me from moving forward? I renamed the main file, renamed the old function, moved the CSS to the SCSS files and deleted the CSS files. I still have to remove the SCSS component imports from themed-bootstrap but do I do that now and then start writing and testing the function? @maxatdetroit

**Comment by maxatdetroit** [comment]
>  I still have to remove the SCSS component imports from themed-bootstrap but do I do that now and then start writing and testing the function?

Yea, that sounds like a good game plan. 

Regarding the CI check that is failing: it's likely due to formatting in the SCSS files you edited. Run `yarn lint` to see where the formatting issue is. See this part of the readme for more instructions: https://github.com/CityOfDetroit/COD-Design-System#formatting-and-linting

**Comment by sreidthomas** [comment]
@maxatdetroit  This is what the script looks like when it runs in the real project:

Screenshot 2024-01-25 at 1 48 33 PM

It converted to css and went into the respective component folder but I also got these errors:

Screenshot 2024-01-25 at 1 50 33 PM

Screenshot 2024-01-25 at 1 50 41 PM

And it added code to the files:

Screenshot 2024-01-25 at 1 52 00 PM

**Comment by maxatdetroit** [comment]
Ah yea, those SCSS files are missing some SCSS mixins that are defined in bootstrap for breakpoints. Previously that wasn't an issue because those SCSS files were being included at the end of the bootstrap base, but now we'll have to put the includes at the top of the individual component SCSS files as well. 

Give me a bit and I'll send you the right includes needed to resolve the errors. Feel free to poke around yourself as well if you want to learn more about bootstrap breakpoints and customizing bootstrap.

**Comment by maxatdetroit** [comment]
@sreidthomas for each SCSS file where you're getting those `undefined variable` or `undefined mixin` errors, make sure these three includes are at the top of the file (in this order):

```scss
@import '../../../../node_modules/bootstrap/scss/functions';
@import '../../../../node_modules/bootstrap/scss/variables';
@import '../../../../node_modules/bootstrap/scss/mixins/breakpoints';
```

Those SCSS libraries include the mixins and variables that are needed to compile that SCSS.

**Comment by sreidthomas** [comment]
I added those lines to code but I am still getting these errors for those files: @maxatdetroit 

Screenshot 2024-01-26 at 10 29 38 AM

The code inside the files:

Screenshot 2024-01-26 at 10 18 54 AM

Screenshot 2024-01-26 at 10 19 08 AM


Screenshot 2024-01-26 at 10 19 18 AM

Screenshot 2024-01-26 at 10 19 28 AM

**Comment by maxatdetroit** [comment]
Would you commit those changes (just the scss files, not the generated files) and push them so I can take a look?

**Comment by sreidthomas** [comment]
> Would you commit those changes (just the scss files, not the generated files) and push them so I can take a look?

Is the security issue fixed that Edgar mentioned @maxatdetroit?

**Comment by sreidthomas** [comment]
I seen storybook test failed but the others passed. I did a merge to try and fix it and now more of the test failed. So I may have made it worse. Not sure if you need those to pass to check things out @maxatdetroit

**Comment by maxatdetroit** [comment]
@sreidthomas , the SCSS files where you're getting those `undefined variable` or `undefined mixin` errors still don't have all those lines I mentioned earlier:

```scss
@import '../../../../node_modules/bootstrap/scss/functions';
@import '../../../../node_modules/bootstrap/scss/variables';
@import '../../../../node_modules/bootstrap/scss/mixins/breakpoints';
```

Make sure those are included in those SCSS files and it should resolve the errors.

The reason you're getting storybook and chromatic test errors on this PR is because:
1. Those tests were added in the `dev` branch which you just recently merged into this branch
2. The generated CSS in this PR is not valid CSS (due to the SASS errors) and so storybook is failing to build and breaking the tests.

Once you fix the SCSS files, then rerun your compilation script, it should fix the generated CSS files and resolve the test failures.

**Comment by sreidthomas** [comment]
Script ran with no issues but the Storybook and Chromatic tests still failed @maxatdetroit

**Comment by maxatdetroit** [comment]
Chromatic is failing because Storybook build is failing. Check out the [build errors on GitHub](https://github.com/CityOfDetroit/COD-Design-System/actions/runs/7699806374/job/20982250183?pr=136) or generate the same error locally with `yarn build`.

What do you think the issue is?

**Comment by maxatdetroit** [comment]
```
ERR! => Failed to build the preview
ERR! Module not found: Error: Can't resolve './TableHeader.css' in '/home/runner/work/COD-Design-System/COD-Design-System/src/components/atoms/TableHeader'
```

It looks like the script is naming the output CSS files like `_tableheader.css`, but the Javascript components are importing `TableHeader.css` (which does not exist). We'll need to update the script to make the output CSS files like `TableHeader.css`. The easiest way to do that will be to rename the source SCSS files like `TableHeader.scss`. That way we shouldn't have to rewrite the script at all.

**Comment by sreidthomas** [comment]
I think I see the issue:
Screenshot 2024-01-29 at 2 51 30 PM

It's looking for a TableCell.css file but that file is being output as: table-cell. Some of the files has a "-" in it. You think I may need to account for it? @maxatdetroit

**Comment by maxatdetroit** [comment]
@sreidthomas yea, definitely. Also the Javascript imports are case sensitive, so make sure the capitalization is correct too. `_table-cell.scss` needs to become `TableCell.scss`, `_tableheadler.scss` needs to become `TableHeader.scss`, etc. That way the correct `css` files will be generated.

**Comment by sreidthomas** [comment]
Still failing. It saying the OffcanvasHeader and OffcanvasBody does not exist. it created all the files with no issue @maxatdetroit 


Screenshot 2024-01-30 at 1 25 09 PM

Screenshot 2024-01-30 at 1 24 53 PM

**Comment by maxatdetroit** [comment]
@sreidthomas check the file names and capitalization. The offcanvas header and body CSS files are still being generated as all lower case: 

image

**Comment by sreidthomas** [comment]
Where are those being generated? I don't see any lowercase names in the directory @maxatdetroit 

Screenshot 2024-01-30 at 2 21 45 PM

**Comment by maxatdetroit** [comment]
If you're still working on your Macbook, then it's important to know that MacOS by default is a 'case insensitive' file system. It's going to consider `offcanvasheader.css` and `OffcanvasHeader.css` as the same file and probably show you just one of them. `git` on the other hand is a case-sensitive source control so it considers `offcanvasheader.css` and `OffcanvasHeader.css` as two different files. 

What's probably happening, is the files got merged together on your OS into a file named `OffcanvasHeader.css` but git is still tracking that file with the name `offcanvasheader.css`. 

You should be able to fix it with a `git mv` to force `git` into tracking the camel case file name (instead of the lower case one):
```bash
git mv offcanvasheader.css OffcanvasHeader.css
```

Sometimes if you're already moved/renamed files quite a few times, then that trick won't work because git is confused. Since these are generated files, it's probably easiest to just deleted the generated file, and regenerate it using your script.

```bash
rm offcanvasheader.css offcanvasheader.css.map
git add offcanvasheader.css offcanvasheader.css.map
# then run your script here to regenerate the files
git add OffcanvasHeader.css OffcanvasHeader.css.map
```

You'd have to do those steps for each file that is still messed up with the case-sensitivity issue.

Read more details here:
* [MacOS is case insensitive](https://discussions.apple.com/thread/251191099?answerId=252281894022&sortBy=best#252281894022)
* [Git is case sensitive](https://learn.microsoft.com/en-us/azure/devops/repos/git/case-sensitivity?view=azure-devops)
* [Fixing files named incorrectly due to case-insensitive file systems](https://stackoverflow.com/questions/17683458/how-do-i-commit-case-sensitive-only-filename-changes-in-git)

**Comment by sreidthomas** [comment]
I tried both commands and it's saying that the file does not exist or that it's fatal. Am I supposed to add the full directory path? @maxatdetroit  

Screenshot 2024-01-31 at 11 04 16 AM

**Comment by maxatdetroit** [comment]
@sreidthomas , whenever you run a command on the command line that takes a filepath as a parameter you have to use a valid path. It can be an absolute path or a relative path, but it must point to the location of a file. 

In the screenshot, a relative path is used but it's not valid because there's no `offcanvasheader.css` in the project root directory. You'd have to specify the directories of the path relative to the project root directory (where you ran the command): `src/components/atoms/offcanvasheader.css`.

**Comment by sreidthomas** [comment]
Okay so I ran the script and it completed with none of the issues I was having a couple of weeks ago. But chromatic and storybook are still failing. Should I try merging COD dev into my branch @maxatdetroit? 
Screenshot 2024-02-21 at 10 55 47 AM

**Comment by maxatdetroit** [comment]
Storybook and chromatic are failing because they can't find the files your script generated. This is still due to the previous [case-sensitive issue we discussed on this PR](https://github.com/CityOfDetroit/COD-Design-System/pull/136#issuecomment-1917813272).

From the [storybook build output for this PR](https://github.com/CityOfDetroit/COD-Design-System/actions/runs/7991693577/job/21823457519?pr=136):
```
ERR! => Failed to build the preview
ERR! Module not found: Error: Can't resolve './OffcanvasHeader.css' in '/home/runner/work/COD-Design-System/COD-Design-System/src/components/atoms/OffcanvasHeader'
ERR! ModuleNotFoundError: Module not found: Error: Can't resolve './OffcanvasHeader.css' in '/home/runner/work/COD-Design-System/COD-Design-System/src/components/atoms/OffcanvasHeader'
```

Then, we can [see that the file `OffcanvasHeader.css` is deleted](https://github.com/CityOfDetroit/COD-Design-System/pull/136/files#diff-81f7c56d21072edfdb32742f37e46a798b9647c0d09074773fe1e4c20c9ee626) in this PR and [a new file of the same name (all lowercase) is introduced](https://github.com/CityOfDetroit/COD-Design-System/pull/136/files#diff-31d46de144e91ccc6e7ac5a3c70094c5523aa3201bbff95489f499c0d6619123).

**Comment by maxatdetroit** [comment]
@sreidthomas let me pull this PR and fix the file name issues and add a commit that you can pull down with the fix.

**Review by maxatdetroit — APPROVED** [review]
@jedgar1mx this LGTM.

**Comment by sreidthomas** [comment]
@maxatdetroit Everything passed except lint & prettier but when I run the command all I get is this:
Screenshot 2024-04-22 at 10 35 40 AM

**Comment by sreidthomas** [comment]
> @maxatdetroit Everything passed except lint & prettier but when I run the command all I get is this: Screenshot 2024-04-22 at 10 35 40 AM

Does that mean it added that line to the file?

**Comment by sreidthomas** [comment]
> > @maxatdetroit Everything passed except lint & prettier but when I run the command all I get is this: Screenshot 2024-04-22 at 10 35 40 AM
> 
> Does that mean it added that line to the file?

Nevermind I think I found the issue @maxatdetroit

**Comment by sreidthomas** [comment]
All ready to go @maxatdetroit @jedgar1mx

File: src/shared/themed-bootstrap.css
Comment (maxatdetroit) [inline src/shared/themed-bootstrap.css]: @sreidthomas this is a remnant of a merge conflict. It needs to be fixed before we can merge in this PR.

**Review by maxatdetroit — CHANGES_REQUESTED** [review]
We missed a conflict from a merge. Otherwise this is good to go.

File: src/shared/themed-bootstrap.css
Comment (maxatdetroit) [reply src/shared/themed-bootstrap.css]: You should be able to fix it by re-running the `compile_styles.sh` script.

Comment (sreidthomas) [reply src/shared/themed-bootstrap.css]: I ran the script again and it keeps failing @maxatdetroit. I tried to switch the command I found in teams but it's still failing. Its not switching rem to em in that file

Screenshot 2024-04-22 at 12 57 09 PM

Comment (sreidthomas) [reply src/shared/themed-bootstrap.css]: Screenshot 2024-04-22 at 1 02 14 PM

Comment (maxatdetroit) [reply src/shared/themed-bootstrap.css]: Try running `yarn install` first. <--- this is always a good step to make sure you have all the right dependencies installed.

**Comment by sreidthomas** [comment]
### **CURRENT ISSUES:**

### **TableCellHeader.css is using em:**

Screenshot 2024-04-22 at 1 45 15 PM

When the script is ran, it checks to convert rem to em in the TableCellHeader file & fails:

Screenshot 2024-04-22 at 1 47 28 PM

However, it is still doing the conversion, just backwards. So its converting the em to rem:

Screenshot 2024-04-22 at 1 48 49 PM

**ISSUE MAY BE THE sed command**

**Comment by maxatdetroit** [comment]
The issue is definitely the sed command @sreidthomas . It looks like the script isn't using the right `sed` command for MacOS. It should be the following (like you shared over Teams chat):

`sed -i -r "s|\([0-9]\)rem|\1em|g" <file_name_variable_here>`

**Comment by sreidthomas** [comment]
> The issue is definitely the sed command @sreidthomas . It looks like the script isn't using the right `sed` command for MacOS. It should be the following (like you shared over Teams chat):
> 
> `sed -i -r "s|\([0-9]\)rem|\1em|g" <file_name_variable_here>`

I tried the one from Teams but it is still doing the same thing and it's only doing it for that file. Also, I noticed it's only converting one of the rules from em to rem. There are at least 3 rules that are em inside that file and they are not being changed @maxatdetroit

**Comment by maxatdetroit** [comment]
Would you push the code you're using so I can take a look?

**Comment by sreidthomas** [comment]
> Would you push the code you're using so I can take a look?

It's basically the same code that you approved in February after we fixed the capitalization issue. I haven't changed anything except merging dev into the branch. The code that's up now should be the same @maxatdetroit

**Comment by sreidthomas** [comment]
> > Would you push the code you're using so I can take a look?
> 
> It's basically the same code that you approved in February after we fixed the capitalization issue. I haven't changed anything except merging dev into the branch. The code that's up now should be the same @maxatdetroit

It didn't cross my mind to check to see what it was doing before I merged DEV until just now.

**Comment by maxatdetroit** [comment]
> It's basically the same code that you approved in February after we fixed the capitalization issue

The code on this PR doesn't have the right sed command to work correctly on MacOS.

**Comment by sreidthomas** [comment]
> > It's basically the same code that you approved in February after we fixed the capitalization issue
> 
> The code on this PR doesn't have the right sed command to work correctly on MacOS.

Yeah, I switched it back to that one before I pushed. When you gave me the new sed command you told me I could use it to get the script to run when testing but to switch back to that one when pushing the branch. I tested with the MAC sed command

**Comment by maxatdetroit** [comment]
Regarding those 3 commits I just made: In #193 I removed all the relative paths to bootstrap from the bootstrap theme SCSS, so I had to come back here and set the `sass` command flag to include the bootstrap node_module folder during compilation so it could find those files.

@jedgar1mx this should be good to merge in now. I tested on my local and chromatic storybook should be good as well.

**Review by maxatdetroit — APPROVED** [review]

**Pull request merged**
